### PR TITLE
Bug 1114453 - [Wifi] Add wifi setting UX and gaia setting for Allow Offf...

### PR DESCRIPTION
**Pull request by tefn3849**
... function.

### CODE REVIEW

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=cbdc6d9dab8b)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e1a5daaf4610)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=fcf99870d3b6)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d4b5eae4c83a)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=15f232c54f60)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f5b8dbc416ba)

File: apps/system/js/wifi.js [hunk 1]
```diff
@@ -191,21 +194,30 @@ var Wifi = {
...
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      var _sleep_mode = false;
+      SettingsListener.observe('wifi.sleep.mode.status', false, function(value)
+      {
+         _sleep_mode = value;
+      });
```
Comment (Seinlin) [inline apps/system/js/wifi.js]: This observe should move to around line 90.
pls ref to wifi.screen_off_timeout

File: apps/system/js/wifi.js [hunk 2]
```diff
@@ -191,21 +194,30 @@ var Wifi = {
...
-        self._alarmId = req.result;
-        releaseCpuLock();
-      };
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      var _sleep_mode = false;
```
Comment (Seinlin) [inline apps/system/js/wifi.js]: I think _sleep_mode is not ncessary.

File: apps/system/js/wifi.js [hunk 3]
```diff
@@ -191,21 +194,30 @@ var Wifi = {
...
+      SettingsListener.observe('wifi.sleep.mode.status', false, function(value)
+      {
+         _sleep_mode = value;
+      });
+      this.wifiSleepMode = _sleep_mode;
```
Comment (Seinlin) [inline apps/system/js/wifi.js]: I think _sleep_mode is not ncessary.

File: apps/system/js/wifi.js [hunk 4]
```diff
@@ -191,21 +194,30 @@ var Wifi = {
...
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      var _sleep_mode = false;
+      SettingsListener.observe('wifi.sleep.mode.status', false, function(value)
+      {
+         _sleep_mode = value;
```
Comment (Seinlin) [inline apps/system/js/wifi.js]: after move, I think you can use 
self.wifiSleepMode = value;

File: apps/system/js/wifi.js [hunk 5]
```diff
@@ -13,6 +13,9 @@ var Wifi = {
   // If it's set to 0, wifi will never be turn off.
   screenOffTimeout: 0,
 
+  // If user wants to allow wifi off.
+  wifiSleepMode: false,
+
   // if Wifi is enabled but disconnected, try to scan for networks every
```
Comment (Seinlin) [inline apps/system/js/wifi.js]: Plese update the coomment here.
// When wifiSleepMode is true,  Wi-Fi will be automatically turned off during sleep to save battery power.

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c146fa32757c)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e68672f5525e)

File: apps/settings/elements/wifi.html [hunk 6]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
         <h2 data-l10n-id="advancedSettings"> Advanced Settings </h2>
       </header>
       <ul>
+        <li id="wifi-sleepMode" class="hint">
+          <label class="pack-switch">
+          <input type="checkbox" name="wifi.sleepMode" data-state="off">
+          <span data-l10n-id="wifi-off-during-sleep">Wi-Fi Off during Sleep</span>
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: Based on the latest l10n guideline, the text context should not be included in the html file.

File: apps/settings/elements/wifi.html [hunk 7]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
...
+          </label>
+	  <p data-l10n-id="wifi-off-during-sleep-description" class="explanation">
+	  When enabled, Wi-Fi will be automatically turned off during sleep to save battery power.
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: Same here.

File: apps/settings/elements/wifi.html [hunk 8]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
         <h2 data-l10n-id="advancedSettings"> Advanced Settings </h2>
       </header>
       <ul>
+        <li id="wifi-sleepMode" class="hint">
+          <label class="pack-switch">
+          <input type="checkbox" name="wifi.sleepMode" data-state="off">
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: nit: Please make an indention.

File: apps/settings/elements/wifi.html [hunk 9]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
         <h2 data-l10n-id="advancedSettings"> Advanced Settings </h2>
       </header>
       <ul>
+        <li id="wifi-sleepMode" class="hint">
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: It is discouraged to use id in settings app to reduce the naming conflicts as all settings panel live in the same dom tree. Please use a class name instead.

File: apps/settings/js/panels/wifi/panel.js [hunk 10]
```diff
@@ -66,6 +67,9 @@ define(function(require) {
         });
 
         // element related events
+	elements.wifiSleepModeStatus.addEventListener('click',
```
Comment (crh0716) [inline apps/settings/js/panels/wifi/panel.js]: indention error.

File: apps/settings/elements/wifi.html (hunk 8 above)
Comment (crh0716) [inline apps/settings/elements/wifi.html]: `data-state="off"` is not required in this case.

File: apps/settings/js/panels/wifi/panel.js [hunk 11]
```diff
@@ -25,6 +25,7 @@ define(function(require) {
           panel: panel,
           wifi: panel,
           wpsColumn: panel.querySelector('.wps-column'),
+          wifiSleepModeStatus: panel.querySelector('#wifi-sleepMode input'),
```
Comment (crh0716) [inline apps/settings/js/panels/wifi/panel.js]: When specifying the "name" attribute with the name of a settings field on an input, settings app binds the settings field to the input automatically, so all changes made to this script are not required.

File: apps/system/js/wifi.js [hunk 12]
```diff
@@ -192,20 +200,23 @@ var Wifi = {
...
-        releaseCpuLock();
-      };
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      // When user wants to allow wifi off then start with a timer, only turn off wifi till timeout.
+      if(this.wifiSleepMode == true)
```
Comment (crh0716) [inline apps/system/js/wifi.js]: Please leave a space after `if` and make `{` in the same line.

File: apps/system/js/wifi.js [hunk 13]
```diff
@@ -192,20 +200,23 @@ var Wifi = {
...
-        self._alarmId = req.result;
-        releaseCpuLock();
-      };
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      // When user wants to allow wifi off then start with a timer, only turn off wifi till timeout.
```
Comment (crh0716) [inline apps/system/js/wifi.js]: This line exceeds the 80 characters limit, please use two line.

File: apps/system/js/wifi.js [hunk 14]
```diff
@@ -13,6 +13,10 @@ var Wifi = {
   // If it's set to 0, wifi will never be turn off.
   screenOffTimeout: 0,
 
+  // When wifiSleepMode is true, Wi-Fi will be automatically turned off
```
Comment (crh0716) [inline apps/system/js/wifi.js]: lint: please remove the extra space in the end.

File: apps/system/js/wifi.js [hunk 15]
```diff
@@ -84,6 +88,10 @@ var Wifi = {
         self.screenOffTimeout = value;
       });
 
+    SettingsListener.observe('wifi.sleepMode', true, function(value){
```
Comment (crh0716) [inline apps/system/js/wifi.js]: lint: please leave a space before `{`.

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6866a332f989)

File: apps/system/js/wifi.js [hunk 16]
```diff
@@ -192,20 +200,23 @@ var Wifi = {
...
-      };
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      // When user wants to allow wifi off then start with a timer,
+      // only turn off wifi till timeout.
+      if(this.wifiSleepMode == true) {
```
Comment (crh0716) [inline apps/system/js/wifi.js]: Space missing before `(`.

File: apps/system/js/wifi.js [hunk 17]
```diff
@@ -84,6 +88,10 @@ var Wifi = {
         self.screenOffTimeout = value;
       });
 
+    SettingsListener.observe('wifi.sleepMode', true, function(value) {
+        self.wifiSleepMode = value;
+     });
```
Comment (crh0716) [inline apps/system/js/wifi.js]: indention error

File: apps/settings/elements/wifi.html [hunk 18]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
         <h2 data-l10n-id="advancedSettings"> Advanced Settings </h2>
       </header>
       <ul>
+        <li class="hint">
+          <label class="pack-switch">
+            <input type="checkbox" name="wifi.sleepMode">
+            <span data-l10n-id="wifi-off-during-sleep">
+	    </span>
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: indention error

File: apps/settings/elements/wifi.html [hunk 19]
```diff
@@ -47,6 +47,15 @@ <h2 data-l10n-id="availableNetworks"> Available Networks </h2>
...
+          </label>
+	  <p data-l10n-id="wifi-off-during-sleep-description" class="explanation">
```
Comment (crh0716) [inline apps/settings/elements/wifi.html]: indention error

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f2a9434fc859)

File: apps/system/js/wifi.js [hunk 20]
```diff
@@ -192,20 +200,23 @@ var Wifi = {
...
-      };
-      req.onerror = function wifi_offAlarmSetFailed() {
-        console.warn('Fail to set wifi sleep timer on Alarm API. ' +
-          'Turn off wifi immediately.');
-        self.sleep();
-        releaseCpuLock();
-      };
+      // When user wants to allow wifi off then start with a timer,
+      // only turn off wifi till timeout.
+      if (this.wifiSleepMode == true) {
```
Comment (alivedise) [inline apps/system/js/wifi.js]: === instead of ==

Comment (alivedise) [inline apps/system/js/wifi.js]: === instead of ==

Comment (alivedise) [inline apps/system/js/wifi.js]:
```js
if (!this.wifiSleepMode) {
  return
}
```

File: apps/system/js/wifi.js [hunk 21]
```diff
@@ -192,20 +200,23 @@ var Wifi = {
...
+          self._alarmId = req.result;
+          releaseCpuLock();
+        };
+        req.onerror = function wifi_offAlarmSetFailed() {
+          console.warn('Fail to set wifi sleep timer on Alarm API. ' +
+            'Turn off wifi immediately.');
+          self.sleep();
+          releaseCpuLock();
+        };
+      }
```
Comment (alivedise) [inline apps/system/js/wifi.js]: Save the brace

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=688f6db18e6a)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6964699bfc6e)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=cf775f64f81a)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=3ae8cc396952)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c0f7c531fa55)

**Comment by try-server-hook** [comment]
[tefn3849 Amy Chung (tefn3849)](https://github.com/tefn3849) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=671186fe5e6c)

**Pull request merged**